### PR TITLE
Require only ActiveRecord, not the entire Rails framework

### PR DESCRIPTION
**Pull request by adomokos**
I was able to require only AR, however, the generators need the `railtie` gem.

Here are the installed gems without `development` and `test`:

gem-dependencies

I wish AR would have its own generator module, but oh well, Rails is a majestic monolith.

### CODE REVIEW

File: lib/rails/sharding.rb [hunk 1]
```diff
@@ -1,4 +1,7 @@
+require 'active_record'
+require 'active_support/core_ext'
 require 'rails'
```
Comment (adomokos) [inline lib/rails/sharding.rb]: This is not the full Rails gem, it's only `railties`.

Comment (hsgubert) [reply lib/rails/sharding.rb]: Not sure I'm following... the `require 'rails'` loads only the railties when you don't have the rails gem installed? Is that it?

**Comment by hsgubert** [comment]
So I was thinking about this problem of removing the Rails dependency... Could we put `rails` as a development dependency, leave only `activerecord` as a runtime dependency and simply ignore the Railties if if rails is not present?

My point is, I don't see a problem in just ignoring Railties and generators when the gem is used outside of Rails. The gem will still work the same way when included in a Rails app. And people that use the gem outside of Rails... well, they won't have the initializer generator and they will have to include the rake tasks manually if they want to.

**Comment by hsgubert** [comment]
About the `Rails.env` issue. I think the best ideia here is to create a `Rails::Sharding.env` method, and use it throughout the gem.

This method just delegates to `Rails.env` if Rails is defined. Otherwise we can create some sort of environment variable like `APP_ENV` or whatever that will set `Rails::Sharding.env`.

It think this approach is cleaner than overriding `Rails.env` somehow and it would semantically decouple the gem environment from the Rails enviroment.

What do you think @adomokos ?

File: lib/rails/sharding.rb (hunk 1 above)
Comment (adomokos) [reply lib/rails/sharding.rb]: That is correct. Instead of using the name `Railties`, they use `Rails`, however, the `rails` gem is not included.

**Comment by adomokos** [comment]
I would like to push the envelop a bit further, if possible. What if we wouldn't even include `railties` or Rails if this gem is not used with Rails. I am thinking about the scenario where we need to fight through Rails originated gem conflicts in a pure Ruby app in the development environment.

What if the gem was "smart enough" to configure itself if Rails is used or not. It's a simple boolean check, I think it could be feasible. We might have to add conditionals to the `.gemspec` file which could be a turn off for some, but the less dependency is the better IMHO.

**Comment by hsgubert** [comment]
I agree with you about the gem being able to work in both "modes" (with and without `rails`), and I think we might achieve this with my earlier suggestion: leave only `activerecord` as a gem runtime dependency and try to include rails in a smart way inside the gem. If rails doesn't exist, just ignore it, and don't register the generators and rake tasks.

I don't know exactly how to do this. I have never "conditionally included" a gem before, but there must be a way to do it.

In this case the gemspec would not need to be conditional at all. As I said earlier, `rails` would be just a development dependency (for gem developers) so we can test the gem in a rails app, but rails would not be installed for your pure ruby app.

**Comment by hsgubert** [comment]
It would be easier for me to make some experiments in that direction if I had a pure ruby app, but I don't. I don't actually even know how to setup one (I just work with rails), but it would be cool if you could try to inplement those "conditional inclusions" of rails and add the gem to your app to see if it works.

**Comment by adomokos** [comment]
Thanks for your comment, I think your suggestion should work. I'll try to make it work when I have a bit of free time. 👍

**Comment by adomokos** [comment]
I am not actively working on this. Closing it.

**Pull request closed without merging**